### PR TITLE
Em 2620 add tags to comments

### PR DESCRIPTION
**Pull request by tomxelliott**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/em/annotation/domain/Tag.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        return color;
+    }
+
+    public void setColor(String color) {
+        this.color = color;
+    }
+
+    public Annotation getAnnotation() {
+        return annotation;
+    }
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/domain/Tag.java]: is this needed? I thought the annotation references the tag and not the other way around. why does the tag need to know about the annotation?

File: src/main/java/uk/gov/hmcts/reform/em/annotation/rest/TagResource.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    @ApiOperation(value = "Get list of tags created by user", notes = "A GET request to retrieve a list of tags")
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "Success", response = AnnotationDTO.class),
+            @ApiResponse(code = 401, message = "Unauthorised"),
+            @ApiResponse(code = 403, message = "Forbidden"),
+            @ApiResponse(code = 404, message = "Not Found"),
+    })
+    @GetMapping("/tags/{createdBy}")
+    //@Timed
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/rest/TagResource.java]: are there no endpoints to add/remove tags?

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/AnnotationServiceImpl.java [hunk 3]
```diff
@@ -64,6 +78,20 @@ public AnnotationDTO save(AnnotationDTO annotationDTO) {
...
         }
+        if (annotationDTO.getTags() != null) {
+            annotation.getTags().forEach(t -> {
+                if (t.getId() == null) {
+                    tagService.persistTag(t);
+                    t.setId(UUID.randomUUID());
+                }
+
+                if (t.getAnnotation() == null) {
+                    t.setAnnotation(annotation);
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/AnnotationServiceImpl.java]: still not sure why we set the annotation on the tag

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/AnnotationServiceImpl.java [hunk 4]
```diff
@@ -64,6 +78,20 @@ public AnnotationDTO save(AnnotationDTO annotationDTO) {
                 }
             });
         }
+        if (annotationDTO.getTags() != null) {
+            annotation.getTags().forEach(t -> {
+                if (t.getId() == null) {
+                    tagService.persistTag(t);
+                    t.setId(UUID.randomUUID());
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/AnnotationServiceImpl.java]: I thought we weren't going to do this, and we were going to save the tags separately

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/dto/TagDTO.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    private String name;
+
+    private String createdBy;
+
+    @Size(max = 20)
+    private String label;
+
+    private String color;
+
+    private UUID annotationId;
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/service/dto/TagDTO.java]: as above, not sure why we need the annotation on the tag

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/CommentServiceImpl.java [hunk 6]
```diff
@@ -28,9 +29,14 @@
...
 
-    public CommentServiceImpl(CommentRepository commentRepository, CommentMapper commentMapper) {
+    private final TagService tagService;
+
+    public CommentServiceImpl(CommentRepository commentRepository,
+                              CommentMapper commentMapper,
+                              TagService tagService) {
         this.commentRepository = commentRepository;
         this.commentMapper = commentMapper;
+        this.tagService = tagService;
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/CommentServiceImpl.java]: is this actually used here?

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/TagServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * @param createdBy the user who owns the tags
+     * @return a list of tags belonging to the required user
+     */
+    @Override
+    public List<TagDTO> findTagByCreatedBy(String createdBy) {
+        return tagRepository.findTagsByCreatedBy(createdBy)
+                .stream()
+                .filter(t -> t.getAnnotation() == null)
+                .filter(distinctByKey(Tag::getName))
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/TagServiceImpl.java]: aren't the tags unique already?

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/TagServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                .map(tagMapper::toDto)
+                .map(tag -> {
+                    tag.setId(null);
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/TagServiceImpl.java]: why do we do this?

File: src/main/java/uk/gov/hmcts/reform/em/annotation/rest/TagResource.java (hunk 2 above)
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/em/annotation/rest/TagResource.java]: It was decided that removal of tags would be an admin task. Adding tags happens as part of the edit comment process.

File: src/main/java/uk/gov/hmcts/reform/em/annotation/domain/Tag.java (hunk 1 above)
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/em/annotation/domain/Tag.java]: It is required for mapping tags to annotations in the database.

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/dto/TagDTO.java (hunk 5 above)
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/em/annotation/service/dto/TagDTO.java]: As above, this is required to map between tags and annotation

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/AnnotationServiceImpl.java (hunk 3 above)
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/em/annotation/service/impl/AnnotationServiceImpl.java]: it is required for mapping

File: src/main/java/uk/gov/hmcts/reform/em/annotation/service/dto/TagDTO.java (hunk 5 above)
Comment (benouaer) [reply src/main/java/uk/gov/hmcts/reform/em/annotation/service/dto/TagDTO.java]: yes, but I thought it's one direction. annotations need to reference tags, but I don't get why tags need to know which annotations they're part of

**Pull request closed without merging**